core/src/main/java/org/elasticsearch/index/mapper/geo/GeoShapeFieldMapper.java

```diff
@@ -257,31 +206,104 @@ public MappedFieldType clone() {
             return new GeoShapeFieldType(this);
         }
 
-        public PrefixTreeStrategy defaultStrategy() {
-            return this.defaultStrategy;
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (!(o instanceof GeoShapeFieldType)) return false;
+            if (!super.equals(o)) return false;
+            GeoShapeFieldType that = (GeoShapeFieldType) o;
+            return Objects.equals(treeLevels, that.treeLevels) &&
+                Objects.equals(precisionInMeters, that.precisionInMeters) &&
+                Objects.equals(defaultDistanceErrorPct, that.defaultDistanceErrorPct) &&
+                Objects.equals(tree, that.tree) &&
+                Objects.equals(strategyName, that.strategyName) &&
+                Objects.equals(distanceErrorPct, that.distanceErrorPct) &&
+                Objects.equals(orientation, that.orientation);
         }
 
-        public PrefixTreeStrategy resolveStrategy(String strategyName) {
-            if (SpatialStrategy.RECURSIVE.getStrategyName().equals(strategyName)) {
-                return recursiveStrategy;
+        @Override
+        public int hashCode() {
+            return Objects.hash(super.hashCode(), tree, strategyName, treeLevels, precisionInMeters, distanceErrorPct, defaultDistanceErrorPct, orientation);
+        }
+
+        @Override
+        public void freeze() {
```

Comment:
I spoke with @jpountz and decided to leave this for now, with a comment on its "hackiness".  The setup in question must happen at the end of field type setup, since this needs the field name to be set. I wanted to keep this setup logic next to the place where the objects are held, instead of outside in the builder. In the future, we could potentially move this to another method and have the builder call it, and then have checks in freeze that these objects have been setup.